[BUG] auth/AuthProvider type hint incorrect

**Checklist**

- [x] I checked the [FAQ section](https://schemathesis.readthedocs.io/en/stable/faq.html#frequently-asked-questions) of the documentation
- [x] I looked for similar issues in the [issue tracker](https://github.com/schemathesis/schemathesis/issues)

**Describe the bug**
I believe the type hint for AuthProvider is incorrect.
The definition of AuthProvider uses a TypeVar for the `data` argument to `get` and `set`.
I think the intended behaviour was to mandate that the types passed to `get` and `set` are the same for a given AuthProvider.



**To Reproduce**
Steps to reproduce the behavior:

1. Implement an Auth Provider (see example below):
```py
from typing import Any

import schemathesis
from schemathesis import Case
from schemathesis.auths import AuthProvider, AuthContext

@schemathesis.auth
class Provider(AuthProvider):
    def get(self, context: AuthContext) -> dict[str, Any]:
        return {"access_token": "foo"}
    def set(self, case: Case, data: dict[str, Any], context: AuthContext) -> None:
        reveal_type(data)
```
2. Run `mypy mwe.py`
3. Get errors about liskov subsitution policy
```
$ mypy mwe.py
mwe.py:11: error: Return type "Dict[str, Any]" of "get" incompatible with return type "Optional[Auth]" in supertype "AuthProvider"  [override]
mwe.py:14: error: Argument 2 of "set" is incompatible with supertype "AuthProvider"; supertype defines the argument type as "Auth"  [override]
mwe.py:14: note: This violates the Liskov substitution principle
mwe.py:14: note: See https://mypy.readthedocs.io/en/stable/common_issues.html#incompatible-overrides
mwe.py:15: note: Revealed type is "builtins.dict[builtins.str, Any]"
Found 2 errors in 1 file (checked 1 source file)
```

**Expected behavior**
No mypy errors

**Environment (please complete the following information):**

- OS: Linux or macOS
- Python version: 3.9.2
- Schemathesis version: 3.17.9
- Spec version: N/A

**Additional context**
If I add `Generic[Auth]` to the `AuthProvider` class definition, the errors go away.

```patch
--- a/src/schemathesis/auths.py
+++ b/src/schemathesis/auths.py
@@ -32,7 +32,7 @@


 @runtime_checkable
-class AuthProvider(Protocol):
+class AuthProvider(Protocol, Generic[Auth]):
     """Get authentication data for an API and set it on the generated test cases."""

     def get(self, context: AuthContext) -> Optional[Auth]:
```

--- a/docs/changelog.rst
+++ b/docs/changelog.rst
@@ -12,6 +12,10 @@ Changelog
 
 - Pin ``Werkzeug`` to ``<3.0``.
 
+**Fixed**
+
+- Fixed type hint for ``AuthProvider``. `#1776_`
+
 .. _v3.19.7:
 
 `3.19.7`_ - 2023-09-03
@@ -3409,6 +3413,7 @@ Deprecated
 .. _0.3.0: https://github.com/schemathesis/schemathesis/compare/v0.2.0...v0.3.0
 .. _0.2.0: https://github.com/schemathesis/schemathesis/compare/v0.1.0...v0.2.0
 
+.. _#1776: https://github.com/schemathesis/schemathesis/issues/1776
 .. _#1763: https://github.com/schemathesis/schemathesis/issues/1763
 .. _#1753: https://github.com/schemathesis/schemathesis/issues/1753
 .. _#1742: https://github.com/schemathesis/schemathesis/issues/1742

--- a/src/schemathesis/auths.py
+++ b/src/schemathesis/auths.py
@@ -32,7 +32,7 @@ class AuthContext:
 
 
 @runtime_checkable
-class AuthProvider(Protocol):
+class AuthProvider(Generic[Auth], Protocol):
     """Get authentication data for an API and set it on the generated test cases."""
 
     def get(self, context: AuthContext) -> Optional[Auth]:

